circuits/src/zk_circuits/valid_wallet_update.rs

```diff
@@ -210,189 +338,64 @@ where
             )?;
+            let withdrawal_applied_to_balance =
+                cs.logic_and(transfer_applies, external_transfer.direction)?;
+            let withdrawal_not_applied = cs.logic_neg(withdrawal_applied_to_balance)?;
+
+            // Either no withdrawal was applied or the fees were zero pre-update
+            let valid_fee_update = cs.logic_or(withdrawal_not_applied, old_fees_are_zero)?;
+            cs.enforce_true(valid_fee_update)?;
+
+            // --- Mint Updates --- //
 
             // Constrain the mints to be set correctly. A valid mint is one of:
-            //  1. The same mint as in the old wallet, if the balance previously existed and
-            //     was not zero'd
-            //  2. The mint of the transfer if the balance was previously zero; balances are
-            //     constrained to be unique elsewhere in the circuit
+            //  1. The same mint as in the old wallet
+            //  2. The old mint was zero and the new mint is the transfer mint
             //  3. The zero mint, if the balance is zero
             let mints_equal = EqGadget::eq(&old_balance.mint, &new_balance.mint, cs)?;
-            let equals_transfer_mint =
+            let old_mint_equals_transfer_mint =
+                EqGadget::eq(&old_balance.mint, &external_transfer.mint, cs)?;
+            let new_mint_equals_transfer_mint =
                 EqGadget::eq(&new_balance.mint, &external_transfer.mint, cs)?;
-            let mint_is_zero = EqZeroGadget::eq_zero(new_balance.mint, cs)?;
-            let new_balance_zero = EqZeroGadget::eq_zero(new_balance.amount, cs)?;
-            let prev_balance_zero = EqZeroGadget::eq_zero(old_balance.amount, cs)?;
+            let old_mint_is_zero = EqZeroGadget::eq_zero(&old_balance.mint, cs)?;
+            let new_mint_is_zero = EqZeroGadget::eq_zero(&new_balance.mint, cs)?;
+            let old_amount_zero = EqZeroGadget::eq_zero(&old_balance.amount, cs)?;
+            let prev_balance_was_zerod = cs.logic_and(old_amount_zero, old_fees_are_zero)?;
 
             // Condition 1 -- same mint as old wallet, balance not zero'd
-            let prev_balance_not_zero = cs.logic_neg(prev_balance_zero)?;
-            let new_balance_not_zero = cs.logic_neg(new_balance_zero)?;
-            let valid_mint1 =
-                cs.logic_and_all(&[mints_equal, prev_balance_not_zero, new_balance_not_zero])?;
+            let valid_mint1 = mints_equal;
 
             // Condition 2 -- new mint added to wallet
-            let valid_mint2 =
-                cs.logic_and_all(&[equals_transfer_mint, prev_balance_zero, new_balance_not_zero])?;
+            let valid_mint2 = cs.logic_and(new_mint_equals_transfer_mint, old_mint_is_zero)?;
 
             // Condition 3 -- withdrawal of entire balance, mint is now zero
-            let valid_mint3 = cs.logic_and(mint_is_zero, new_balance_zero)?;
+            let valid_mint3 = cs.logic_and(old_mint_equals_transfer_mint, new_mint_is_zero)?;
```

Comment:
similar thing here, the invariant that zero-mint balances have zeros in every other fields means we don't have to check the amount here